### PR TITLE
An example change of unfolding **options for public API classes

### PR DESCRIPTION
**Pull request by Pyifan**
- List the config options of the class (excl. those inherited from parents) in __init__ method
- Move documentation from config class to entity class
- Note that __init__ param of None value will be filtered out before passing to parent class __init__ for building options, so they will take the default value defined in ConfigOption
- Some format correction for docstring

### CODE REVIEW

File: testplan/runners/pools/remote.py [hunk 1]
```diff
@@ -577,19 +579,70 @@ class RemotePoolConfig(PoolConfig):
...
+    default_workspace_root = workspace_root()
+
+    @classmethod
+    def get_options(cls):
+        """
+        Schema for options validation and assignment of default values.
+        """
+        return {
+            'hosts': dict,
+            ConfigOption('abort_signals', default=[signal.SIGINT,
```
Comment (ryan-collingham) [inline testplan/runners/pools/remote.py]: I'm not sure if we need to use ConfigOption at all if we define the default value in the class constructor?

Doesn't need changing for this PR but something to bear in mind in future, I think we can potentially reduce our use of ConfigOption quite a bit.

File: tests/functional/testplan/test_timeout.py [hunk 2]
```diff
@@ -62,3 +62,4 @@ def test_runner_timeout():
         assert current_proc.children() == start_procs
     finally:
         os.remove(output_json)
+        pass
```
Comment (ryan-collingham) [inline tests/functional/testplan/test_timeout.py]: Why is this needed...?

File: tests/unit/testplan/common/config/test_generic_config.py [hunk 3]
```diff
@@ -157,3 +158,95 @@ def test_getattr_propagation():
...
+
+    @property
+    def options(self):
+        return self._options
+
+
+def test_filter_locals():
+    """Test that Entity.filter_locals() works correctly."""
+    bottom1 = Bottom('Bottom1', baz='you', bar='bye', boo=None)
+    # Arguments defined explicitly in __init__() should appear and for mutable
```
Comment (ryan-collingham) [inline tests/unit/testplan/common/config/test_generic_config.py]: So what's the expected behaviour for mutable types? If I append 4 to bottom1.cfg.boo does that mean if I create a new Bottom instance its boo will default to [1, 2, 3, 4] or just [1, 2, 3]? I think the former would be very confusing.

File: tests/unit/testplan/common/config/test_generic_config.py [hunk 4]
```diff
@@ -157,3 +158,95 @@ def test_getattr_propagation():
...
+    assert bottom1.options['baz'] == 'you'
+    assert bottom1.options['bar'] == 'bye'
+    assert bottom1.options['foo'] == 9
+
+    # takes the value of immediate container
+    assert bottom1.cfg.boo == [1, 2, 3]
+
+    bottom2 = Bottom(
+        'Bottom2', description='An example',
+        boo=None, zoo={10: 'a', 20: 'b', 30: 'c'}, bar='barbar'
```
Comment (ryan-collingham) [inline tests/unit/testplan/common/config/test_generic_config.py]: So what does bottom2.cfg.boo get set to - None or [1, 2, 3]...?

File: testplan/runners/pools/remote.py (hunk 1 above)
Comment (Pyifan) [reply testplan/runners/pools/remote.py]: i think this is for differentiating optional and mandatory options

File: tests/unit/testplan/common/config/test_generic_config.py (hunk 3 above)
Comment (Pyifan) [reply tests/unit/testplan/common/config/test_generic_config.py]: hmmm, maybe the comment is just misleading. For init param that has a None default, they are going to take the default value defined in its ConfigOption. And this is the recommended usage for options that has mutable type of default value.

File: tests/unit/testplan/common/config/test_generic_config.py (hunk 4 above)
Comment (Pyifan) [reply tests/unit/testplan/common/config/test_generic_config.py]: [1, 2, 3] from immediate container

File: tests/functional/testplan/test_timeout.py (hunk 2 above)
Comment (Pyifan) [reply tests/functional/testplan/test_timeout.py]: probably added for debugging purpose, removed.

File: testplan/runners/pools/remote.py (hunk 1 above)
Comment (ryan-collingham) [reply testplan/runners/pools/remote.py]: We don't need that differentiation when we are unfolding **options and providing a default value in the constructor.
